lib/transport-common.js

```diff
@@ -5,6 +5,30 @@ const Application = require('./applications').Application;
 const SimpleConnectPolicy = require('./simple-connect-policy');
 const common = require('./common');
 
+const DEFAULT_BACKOFF_TIME = 1000;
+const MAX_BACKOFF_TIME = 32000;
+const RANDOM_OFFSET_MAX = 1000;
+
+function createDefaultReconnector() {
+  let backoff = DEFAULT_BACKOFF_TIME;
+  const randomMsecGen = () =>
+    Math.floor((RANDOM_OFFSET_MAX + 1) * common.cryptoRandom());
+  return (connection, reconnectFn) => {
+    if (connection.closedIntentionally) return;
+    setTimeout(() => {
```

Comment:
Do I understand it correctly that default reconnector will try to reconnect indefinitely after reaching `MAX_BACKOFF_TIME` with `MAX_BACKOFF_TIME` in case there is always an error or have I missed something? If it's the case I think it'd be better to stop after trying with `MAX_BACKOFF_TIME` once or at least make amount of retries configurable.

Reply:
You can call `connection.close` to stop reconnection with this default reconnector.
Also, users of the library can always provide their own reconnector implementation that may behave like you described or that may have any other different logic behind it.
I don't see how we can make the number of retries configurable (since it is the default reconnector, used when none is provided) and I don't think that it is okay to just let us make process termination (when reconnection is the only thing keeping it alive) default behavior if the client isn't able to connect successfully during the period of ~70 seconds.

Reply:
> I don't see how we can make the number of retries configurable

Oh, I thought that we expose this function to the user, that's why I suggested it. I think we should expose it as backoff is a common strategy. This way we can just add an argument that will control this behavior.

>  I don't think that it is okay to just let us make process termination default behavior

But this process will only use user resources without any utility to the user and the only way to stop it will be `kill`. I don't think it's appropriate to keep alive useless process without explicit approval of the user. Which also adds to the point of just exposing this function and giving user the control. As for the default behavior I think it's better to just stop trying but maybe @nechaido  can add smth?

Reply:
>But this process will only use user resources without any utility to the user and the only way to stop it will be `kill`.

I mean, retries are asynchronous, `kill` is not the only option to terminate the process.
A user can always use `connection.close` to stop reconnection, this was initially @nechaido's idea.

>to keep alive useless process without explicit approval of the user

What do you mean by 'useless'? It is not useless, if the process is supposed to communicate with the server after the connection is established, then it is still the task of this process to keep the connection alive and not just terminate the process when the connection is lost for a minute, possibly losing the data. A client can even make some calls when the connection is lost, and they will be transmitted to the server when it reconnects successfully, and if the client doesn't want it to happen, it can use `connection.close` method explicitly closing the connection.

I am okay with exposing the function, but in that case, we need to decide on what things will be configurable and we should also create some other alternatives, like the one disabling reconnection, for example (possibly making it the default one, to avoid changing the behavior). Also, I am not sure about how it should behave with this argument specifying the number of retries, is it a total count of retries since the first, or is it the count of retries since the first time maximum backoff time was used?

Reply:
I am okay with exposing the function, in case we make it configurable, but I would prefer that its default behaviour stays the same.